### ISSUE TITLE
Stream loses values when using different dispatchers

### ISSUE THREAD

**Issue by christian-laakmann** [reactor/reactor#479]
Hi all,

I'm trying to create a stream that parallellizes two steps of my stream pipeline and then continues the process in single threaded mode.
However, on my machine, the test case below only works around ~10% of the time. Most executions end with about 100000 events not being registered in the CountDownLatch.

For some reasons, the test completes on a colleague's machine with the same specs...

Am I doing something wrong here? Any ideas?
See below for a simplified test case and my console output with thread dump.

Thanks in advance

```
  @Test
  public void testStream() throws InterruptedException {

    Environment environment = Environment.initializeIfEmpty();

    int end = 1000000;
    final CountDownLatch latch = new CountDownLatch(end + 1);

    Consumer<String> action = name -> {};

    DispatcherSupplier supplier = Environment.newCachedDispatchers(20, "groupByPool");

    new Thread(() -> {
      Streams.range(0, end)
             .dispatchOn(environment, supplier.get())
             .map(Long::intValue)
             .map(Object::toString)
             .observe(i -> action.accept("stage-1"))
             .observe(i -> action.accept("stage-2"))
             .partition()
             .flatMap(stream -> stream.dispatchOn(environment, supplier.get())
                                      .observe(i -> action.accept("stage-3-p"))
                                      .observe(i -> action.accept("stage-4-p"))
                      //.log("partition")
             )
             .dispatchOn(environment, supplier.get())
             .observe(i -> action.accept("stage-5"))
             .observe(i -> action.accept("stage-6"))
              .observe(i -> latch.countDown())
                      //.log("default")
              .consume();
    }).start();

    if(!latch.await(20, TimeUnit.SECONDS)) {
      Thread.getAllStackTraces().forEach((thread, trace) -> {

        String output = thread.getName() + ": " + "\n";
        for (StackTraceElement stackTraceElement : trace) {
          output += "\t" + stackTraceElement.toString() + "\n";
        }

        System.out.println(output + "\n");

      });

      throw new IllegalStateException(latch.getCount() + " pending events");
    }

    assertEquals(0, latch.getCount());
  }
```

Output:

```
groupByPool-10: 
	sun.misc.Unsafe.park(Native Method)
	java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	reactor.jarjar.com.lmax.disruptor.BlockingWaitStrategy.waitFor(BlockingWaitStrategy.java:45)
	reactor.core.dispatch.wait.AgileWaitingStrategy.waitFor(AgileWaitingStrategy.java:47)
	reactor.jarjar.com.lmax.disruptor.ProcessingSequenceBarrier.waitFor(ProcessingSequenceBarrier.java:55)
	reactor.jarjar.com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:123)
	java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	java.lang.Thread.run(Thread.java:745)


groupByPool-3: 
	sun.misc.Unsafe.park(Native Method)
	java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	reactor.jarjar.com.lmax.disruptor.BlockingWaitStrategy.waitFor(BlockingWaitStrategy.java:45)
	reactor.core.dispatch.wait.AgileWaitingStrategy.waitFor(AgileWaitingStrategy.java:47)
	reactor.jarjar.com.lmax.disruptor.ProcessingSequenceBarrier.waitFor(ProcessingSequenceBarrier.java:55)
	reactor.jarjar.com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:123)
	java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	java.lang.Thread.run(Thread.java:745)


Finalizer: 
	java.lang.Object.wait(Native Method)
	java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:142)
	java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:158)
	java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:209)


groupByPool-4: 
	sun.misc.Unsafe.park(Native Method)
	java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	reactor.jarjar.com.lmax.disruptor.BlockingWaitStrategy.waitFor(BlockingWaitStrategy.java:45)
	reactor.core.dispatch.wait.AgileWaitingStrategy.waitFor(AgileWaitingStrategy.java:47)
	reactor.jarjar.com.lmax.disruptor.ProcessingSequenceBarrier.waitFor(ProcessingSequenceBarrier.java:55)
	reactor.jarjar.com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:123)
	java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	java.lang.Thread.run(Thread.java:745)


groupByPool-8: 
	sun.misc.Unsafe.park(Native Method)
	java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	reactor.jarjar.com.lmax.disruptor.BlockingWaitStrategy.waitFor(BlockingWaitStrategy.java:45)
	reactor.core.dispatch.wait.AgileWaitingStrategy.waitFor(AgileWaitingStrategy.java:47)
	reactor.jarjar.com.lmax.disruptor.ProcessingSequenceBarrier.waitFor(ProcessingSequenceBarrier.java:55)
	reactor.jarjar.com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:123)
	java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	java.lang.Thread.run(Thread.java:745)


groupByPool-6: 
	sun.misc.Unsafe.park(Native Method)
	java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	reactor.jarjar.com.lmax.disruptor.BlockingWaitStrategy.waitFor(BlockingWaitStrategy.java:45)
	reactor.core.dispatch.wait.AgileWaitingStrategy.waitFor(AgileWaitingStrategy.java:47)
	reactor.jarjar.com.lmax.disruptor.ProcessingSequenceBarrier.waitFor(ProcessingSequenceBarrier.java:55)
	reactor.jarjar.com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:123)
	java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	java.lang.Thread.run(Thread.java:745)


groupByPool-7: 
	sun.misc.Unsafe.park(Native Method)
	java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	reactor.jarjar.com.lmax.disruptor.BlockingWaitStrategy.waitFor(BlockingWaitStrategy.java:45)
	reactor.core.dispatch.wait.AgileWaitingStrategy.waitFor(AgileWaitingStrategy.java:47)
	reactor.jarjar.com.lmax.disruptor.ProcessingSequenceBarrier.waitFor(ProcessingSequenceBarrier.java:55)
	reactor.jarjar.com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:123)
	java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	java.lang.Thread.run(Thread.java:745)


groupByPool-1: 
	sun.misc.Unsafe.park(Native Method)
	java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	reactor.jarjar.com.lmax.disruptor.BlockingWaitStrategy.waitFor(BlockingWaitStrategy.java:45)
	reactor.core.dispatch.wait.AgileWaitingStrategy.waitFor(AgileWaitingStrategy.java:47)
	reactor.jarjar.com.lmax.disruptor.ProcessingSequenceBarrier.waitFor(ProcessingSequenceBarrier.java:55)
	reactor.jarjar.com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:123)
	java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	java.lang.Thread.run(Thread.java:745)


Monitor Ctrl-Break: 
	java.net.DualStackPlainSocketImpl.accept0(Native Method)
	java.net.DualStackPlainSocketImpl.socketAccept(DualStackPlainSocketImpl.java:131)
	java.net.AbstractPlainSocketImpl.accept(AbstractPlainSocketImpl.java:404)
	java.net.PlainSocketImpl.accept(PlainSocketImpl.java:199)
	java.net.ServerSocket.implAccept(ServerSocket.java:545)
	java.net.ServerSocket.accept(ServerSocket.java:513)
	com.intellij.rt.execution.application.AppMain$1.run(AppMain.java:90)
	java.lang.Thread.run(Thread.java:745)


Reference Handler: 
	java.lang.Object.wait(Native Method)
	java.lang.Object.wait(Object.java:502)
	java.lang.ref.Reference$ReferenceHandler.run(Reference.java:157)


Signal Dispatcher: 


main: 
	java.lang.Thread.dumpThreads(Native Method)
	java.lang.Thread.getAllStackTraces(Thread.java:1603)
	com.novomind.iport.process2.ProcessTest.testStream(ProcessTest.java:227)
	sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	java.lang.reflect.Method.invoke(Method.java:483)
	org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	mockit.integration.junit4.internal.JUnit4TestRunnerDecorator.executeTestMethod(JUnit4TestRunnerDecorator.java:131)
	mockit.integration.junit4.internal.JUnit4TestRunnerDecorator.invokeExplosively(JUnit4TestRunnerDecorator.java:71)
	mockit.integration.junit4.internal.MockFrameworkMethod.invokeExplosively(MockFrameworkMethod.java:42)
	sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	java.lang.reflect.Method.invoke(Method.java:483)
	mockit.internal.util.MethodReflection.invokeWithCheckedThrows(MethodReflection.java:112)
	mockit.internal.mockups.MockMethodBridge.callMock(MockMethodBridge.java:85)
	mockit.internal.mockups.MockMethodBridge.invoke(MockMethodBridge.java:44)
	org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java)
	org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:78)
	com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:212)
	com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)
	sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	java.lang.reflect.Method.invoke(Method.java:483)
	com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)


Attach Listener: 


groupByPool-9: 
	sun.misc.Unsafe.park(Native Method)
	java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	reactor.jarjar.com.lmax.disruptor.BlockingWaitStrategy.waitFor(BlockingWaitStrategy.java:45)
	reactor.core.dispatch.wait.AgileWaitingStrategy.waitFor(AgileWaitingStrategy.java:47)
	reactor.jarjar.com.lmax.disruptor.ProcessingSequenceBarrier.waitFor(ProcessingSequenceBarrier.java:55)
	reactor.jarjar.com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:123)
	java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	java.lang.Thread.run(Thread.java:745)


groupByPool-5: 
	sun.misc.Unsafe.park(Native Method)
	java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	reactor.jarjar.com.lmax.disruptor.BlockingWaitStrategy.waitFor(BlockingWaitStrategy.java:45)
	reactor.core.dispatch.wait.AgileWaitingStrategy.waitFor(AgileWaitingStrategy.java:47)
	reactor.jarjar.com.lmax.disruptor.ProcessingSequenceBarrier.waitFor(ProcessingSequenceBarrier.java:55)
	reactor.jarjar.com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:123)
	java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	java.lang.Thread.run(Thread.java:745)


groupByPool-2: 
	sun.misc.Unsafe.park(Native Method)
	java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	reactor.jarjar.com.lmax.disruptor.BlockingWaitStrategy.waitFor(BlockingWaitStrategy.java:45)
	reactor.core.dispatch.wait.AgileWaitingStrategy.waitFor(AgileWaitingStrategy.java:47)
	reactor.jarjar.com.lmax.disruptor.ProcessingSequenceBarrier.waitFor(ProcessingSequenceBarrier.java:55)
	reactor.jarjar.com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:123)
	java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	java.lang.Thread.run(Thread.java:745)



java.lang.IllegalStateException: 88278 pending events
	at com.novomind.iport.process2.ProcessTest.testStream(ProcessTest.java:238)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:78)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:212)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
```

**Comment by jbrisbin** [comment]
I think you might be terminating your test thread too early. 

If I put a wait inside the Runnable that waits on the latch to be counted down to ensure everything’s been processed, the test completes normally. If I remove it, then the Thread is exiting before all the work in the Stream has been completed because there’s no way for the Thread to know there’s still work outstanding (it’s not a daemon thread).

~~~java
		new Thread(() -> {
			Streams.range(0, end)
			       .dispatchOn(Environment.get(), supplier.get())
			       .map(Long::intValue)
			       .map(Object::toString)
			       .observe(i -> action.accept("stage-1"))
			       .observe(i -> action.accept("stage-2"))
			       .partition()
			       .flatMap(stream -> stream.dispatchOn(Environment.get(), supplier.get())
			                                .observe(i -> action.accept("stage-3-p"))
			                                .observe(i -> action.accept("stage-4-p"))
			       )
			       .dispatchOn(Environment.get(), supplier.get())
			       .observe(i -> action.accept("stage-5"))
			       .observe(i -> action.accept("stage-6"))
			       .observe(i -> latch.countDown())
			       .consume();
			try {
				latch.await();
			} catch (InterruptedException e) {
			}
		}).start();
~~~

**Comment by smaldini** [comment]
We should provide more options regarding daemon actually on Environment init, a few people got tricked by that unfortunately, that's why the processor threads are non daemon by default (I think we should envision and document that in a future near we will make dispatchers depend on processor meaning they will be non daemon by default).

**Comment by jbrisbin** [comment]
It would also be good to rework this example to use a `Processor` instead of a `Dispatcher` since that's the direction we're going for almost everything (even dispatchers will eventually just delegate to a `Processor`).

**Comment by christian-laakmann** [comment]
@jbrisin I thought I tried that already, but I'll have a look again tomorrow. 

I'd very much like to go with the direction that reactor is taking. How would this look with Processors instead? 

BTW, thanks for this great framework, it's helped us tremendously to get the speed we needed for our application!

**Comment by christian-laakmann** [comment]
I'm still having the same problems, even when I include `latch.await()` in the test thread :(

Any further ideas?

**Comment by smaldini** [comment]
That's strange I will be able to test that post 2.0.3 (unfortunately) that needs to be released for our friends at Spring XD and Spring Integration Kafka.

Can you verify this works for you in the meantime ? (increasing the input range ofc)
```java
	DispatcherSupplier supplier1 = Environment.newCachedDispatchers(2, "groupByPool");
		DispatcherSupplier supplier2 = Environment.newCachedDispatchers(5, "partitionPool");

		CountDownLatch latch = new CountDownLatch(10);

		Streams
				.range(1, 10)
				.groupBy(n -> n % 2 == 0)
				.flatMap(stream -> stream
								.dispatchOn(supplier1.get())
								.log("groupBy")
				)
				.partition(5)
				.flatMap(stream -> stream
								.dispatchOn(supplier2.get())
								.log("partition")
				)
				.dispatchOn(Environment.sharedDispatcher())
				.log("join")
				.consume(t -> {
					latch.countDown();
				});


		assertThat("Not totally dispatched", latch.await(30, TimeUnit.SECONDS));
```

**Comment by christian-laakmann** [comment]
Hi @smaldini,

I'm having the same problems with your test, when I increase the input range.

I have slightly modified the test to find out, at what input size the test fails:

```
    DispatcherSupplier supplier1 = Environment.newCachedDispatchers(2, "groupByPool");
    DispatcherSupplier supplier2 = Environment.newCachedDispatchers(5, "partitionPool");

    for(int count = 1000; count < 1000000; count += 1000){
      CountDownLatch latch = new CountDownLatch(count);
      long start = System.nanoTime();

      Streams
              .range(1, count)
              .groupBy(n -> n % 2 == 0)
              .flatMap(stream -> stream
                               .dispatchOn(supplier1.get())
                               .log("groupBy")
              )
              .partition(5)
              .flatMap(stream -> stream
                               .dispatchOn(supplier2.get())
                               .log("partition")
              )
              .dispatchOn(Environment.sharedDispatcher())
              .log("join")
              .consume(t -> {
                latch.countDown();
              },
                       Throwable::printStackTrace,
                       v -> {
                         long end = System.nanoTime();

                         System.out.println("Stream completed after " + TimeUnit.NANOSECONDS.toSeconds(end - start) + " seconds");
                       });

      if(!latch.await(30, TimeUnit.SECONDS)) {
        System.out.println("COUNT = " + count + ", waiting for " + latch.getCount() + " events");
        assertEquals("Not totally dispatched", 0, latch.getCount());
      }
    }
```

It usually ends with something in the upper ten-thousand range, and as you can see, the stream had plenty time left to complete all inputs:

```
14:38:13.061 [shared-1] INFO  join - onNext: 16982
14:38:13.061 [shared-1] INFO  join - onNext: 16997
14:38:13.061 [shared-1] INFO  join - onNext: 16992
14:38:13.061 [shared-1] INFO  join - complete: DispatcherAction
14:38:13.061 [shared-1] INFO  join - cancel: DispatcherAction
Stream completed after 3 seconds
COUNT = 17000, waiting for 67 events

java.lang.AssertionError: Not totally dispatched 
Expected :0
Actual   :67
 <Click to see difference>
	at com.novomind.iport.process2.ProcessTest.testStream2(ProcessTest.java:292)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:78)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:212)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)

```

**Comment by smaldini** [comment]
All right that's critical, which reactor version and JVM are you running in case ? 

The alternative path is to use http://projectreactor.io/docs/reference/#work for scaling up (at even more speed). You just create the workProcessor and subscribe to it as many time you want threads, e.g:

```java
Processor<String, String> processor = RingBufferWorkProcessor.create("thread-name");
Streams<String> stream = Streams.wrap(processor);

//attach 8 consumers, so 8 threads
for(int i = 0; i < 8; i++){
  stream.consume(System.out::println);
}

processor.onNext("test!");
//or
//Streams.range(1,100000).map(Object::toString).subscribe(processor);
//or
//Stream<String> stream = Streams.range(1,100000).map(Object::toString).process(processor);
//for(int i = 0; i < 8; i++) stream.consume(consumer);
```

**Comment by christian-laakmann** [comment]
JDK 1.8.0_25
Reactor 2.0.2.RELEASE

If you need any further information, I'll be available on your Gitter the next 1 1/2 hours.

Thanks for the processor info, I will check that out.

**Comment by christian-laakmann** [comment]
FYI, I am still having this issue with the latest snapshot, using the test cases above. At some iteration, not all items are processed.

Most of our implementation has been moved to RingBufferWorkProcesser - where we have some issues with correct completion -> I'll create a new issue

**Comment by smaldini** [comment]
Will retest then thanks for the feedback
